### ISSUE TITLE
Override __init__ prototype

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1062]
Hello,

Using autodoc, I find usefull to override functions prototypes but it fails for ****init**** method.

With the following module, the prototype of the **Foo** class is **Foo([message])**, but **show_foo** is **show_foo()**.

Regards

```
#!python
# -*- coding: utf-8 -*-

"""Module to test

"""

from os.path import isfile

class Foo(object):
    """Test class to display foo
    """

    def __init__(self, message='Foo'):
        """__init__()

        Initialize Foo
        """
        self.message = message

    def show_foo(message='foo'):
        """show_foo()

        Display foo
        """
        print message


```

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1062
- Originally reported by: Anonymous
- Originally created at: 2012-12-18T17:20:10.508

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2013-01-20 02:01:15+00:00_

I recommend that you write the following.

```
#!python

    def show_foo(self, message='foo'):
        """
        Display foo
        """
        print message
```

**Comment by shimizukawa** [comment]
_From Andrew Dawson on 2013-01-23 14:13:44+00:00_

I think this has been misunderstood. The question is why can one not override the signature of a class constructor?

Consider:

```
#!python

class MyClass(object):
    "Some class."""

    def __init__(self, *args, **kwargs):
        """
        __init__(arg1, arg2, keyword1=True, keyword2=False)

        Initialize a MyClass instance, a wrapper around OtherClass.

        """
        self.other = OtherClass(*args, **kwargs)

    def my_method(self, *args, **kwargs):
        """
        my_method(arg1, arg2, keyword1='keyword1', keyword2=None)

        Call a method.

        """
        self.other.some_method(*args, **kwargs)

```

Currently it is not possible (as far as I can tell) to override the signature of the constructor, however, it is possible to override the signature of a method. In the above example the signature for the constructor will be:

```
#!python

__init__(*args, **kwargs)
```

but the signature for the method will be:

```
#!python

my_method(arg1, arg2, keyword1='keyword1', keyword2=None)
```

which is I think what was being suggested by the original post.

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2013-01-24 09:05:37+00:00_

OK, I have misunderstood. I'm going to check why.

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2013-01-26 01:47:54+00:00_

[Andrew Dawson](https://bitbucket.org/ajdawson): I wrote a patch.
https://bitbucket.org/shimizukawa/sphinx-init-signature-fork/commits/c074825846ec2ad0663d52910ad35125654965c0

Make sure that change satisfies your requirement and there are not side effects.

**Comment by shimizukawa** [comment]
_From Andrew Dawson on 2013-01-28 09:30:19+00:00_

[Takayuki Shimizukawa](https://bitbucket.org/shimizukawa) This seems to work for me, and I haven't noticed any obvious side-effects.

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2013-02-03 11:31:09+00:00_

Closes #1062: sphinx.ext.autodoc use **init** method signature for class signature.

→ <<cset e70a43a9c53759d4b4b8f5109a02fe0e793835ff>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2013-02-04 00:02:40+00:00_

I implemented code different from before because I found problem by test.
